# Implement A Neural Network From Scratch

In [3]:
import numpy as np
import time
from tensorflow.examples.tutorials.mnist import input_data

mnist = input_data.read_data_sets("Mnist_data/", one_hot=True)
IMGS = mnist.train.images
LABS = mnist.train.labels
IMGS_Test = mnist.test.images
LABS_Test = mnist.test.labels

def softmax(z):
    return np.exp(z) / np.sum(np.exp(z))


def softmax_prime(z):
    return softmax(z) * (1 - softmax(z))

def relu(z):
    return np.maximum(z, 0)

# def relu_prime(z):
#     result = []
#     for t in z:
#         result.append(float(t > 0))
#     return np.array(result, ndmin=2).T

# This one is a lot faster
def relu_prime(z):
    x = np.array(z)
    return np.ceil((abs(x)+x)/2/100)

epochs = 20
lr = 0.2
batch_size = 64
n = mnist.train.num_examples
total_batch = int(n/batch_size)
n_step = epochs * total_batch

beta_1 = 0.9
beta_2 = 0.999
epsilon = 0.00000001

n_input = 784
n_hidden = 100
n_output = 10

w1 = np.random.normal(0, 0.1 ,(n_hidden, n_input))
b1 = np.random.normal(0, 0.1 ,(n_hidden, 1))
w2 = np.random.normal(0, 0.1 ,(n_output, n_hidden))
b2 = np.random.normal(0, 0.1 ,(n_output, 1))

for j in range(epochs):
    time_start = time.time()
    error = 0
    error_t = 0
    cost = 0
    for k in range(0,n,batch_size):
        imgs = IMGS[k:k+batch_size]
        labels = LABS[k:k+batch_size]
        imgs_test = IMGS_Test
        labels_test = LABS_Test
        w1_gd = 0
        w2_gd = 0
        b1_gd = 0
        b2_gd = 0
        t = 0
        a2_m = 0
        a1_m = 0
        a2_v = 0
        a1_v = 0
        a2_m_b = 0
        a1_m_b = 0
        a2_v_b = 0
        a1_v_b = 0

        for x, y in zip(imgs, labels):
            x = np.array(x, ndmin=2).T
            y = np.array(y, ndmin=2).T
            z1 = np.dot(w1, x) + b1
            a1 = relu(z1)
            z2 = np.dot(w2, a1) + b2
            a2 = softmax(z2)

            d2 = a2 - y
            d1 = relu_prime(z1)*np.dot(w2.T, d2)

            w1_gd += np.dot(d1, x.T)
            w2_gd += np.dot(d2, a1.T)
            b1_gd += d1
            b2_gd += d2
            error += int(np.argmax(a2) != np.argmax(y))
            cost = np.sum(y*np.log(a2) + (1-y)*np.log(1-a2))
#             # Adam
#             t += 1
#             a2_m = a2_m * beta_1 + (1-beta_1)*w2_gd
#             a1_m = a1_m * beta_1 + (1-beta_1)*w1_gd
#             a2_m_b = a2_m_b * beta_1 + (1-beta_1)*b2_gd
#             a1_m_b = a1_m_b * beta_1 + (1-beta_1)*b1_gd
        
#             a2_v = a2_v * beta_2 + (1-beta_2)* (w2_gd ** 2)
#             a1_v = a1_v * beta_2 + (1-beta_2)* (w1_gd ** 2)
#             a2_v_b = a2_v_b * beta_2 + (1-beta_2)* (b2_gd ** 2)
#             a1_v_b = a1_v_b * beta_2 + (1-beta_2)* (b1_gd ** 2)
        
#             a2_m_nor = a2_m / (1 - (beta_1 ** t))
#             a1_m_nor = a1_m / (1 - (beta_1 ** t))
#             a2_v_nor = a2_v / (1 - (beta_2 ** t))
#             a1_v_nor = a1_v / (1 - (beta_2 ** t))
#             a2_m_b_nor = a2_m_b / (1 - (beta_1 ** t))
#             a1_m_b_nor = a1_m_b / (1 - (beta_1 ** t))
#             a2_v_b_nor = a2_v_b / (1 - (beta_2 ** t))
#             a1_v_b_nor = a1_v_b / (1 - (beta_2 ** t))        
#         w2 -= (a2_m_nor / (np.sqrt(a2_v_nor) + epsilon)) / batch_size
#         w1 -= (a1_m_nor / (np.sqrt(a1_v_nor) + epsilon)) / batch_size
#         b2 -= (a2_m_b_nor / (np.sqrt(a2_v_b_nor) + epsilon)) / batch_size
#         b1 -= (a1_m_b_nor / (np.sqrt(a1_v_b_nor) + epsilon)) / batch_size
        
        # Gradient Descent
        w1 -= lr * w1_gd / batch_size
        w2 -= lr * w2_gd / batch_size
        b1 -= lr * b1_gd / batch_size
        b2 -= lr * b2_gd / batch_size
    time_end = time.time()
    print("Epoch: "+str(j+1)+"\tCost: {:.3f}\t".format(-cost)+"\tAccuracy: {:.3f}\t".format(1 - error/n)+
          "\tTime: {:.3f}\t".format(time_end-time_start))

Extracting Mnist_data/train-images-idx3-ubyte.gz
Extracting Mnist_data/train-labels-idx1-ubyte.gz
Extracting Mnist_data/t10k-images-idx3-ubyte.gz
Extracting Mnist_data/t10k-labels-idx1-ubyte.gz
Epoch: 1	Cost: 0.332		Accuracy: 0.902		Time: 8.574	
Epoch: 2	Cost: 0.077		Accuracy: 0.953		Time: 8.662	
Epoch: 3	Cost: 0.048		Accuracy: 0.966		Time: 9.565	
Epoch: 4	Cost: 0.041		Accuracy: 0.973		Time: 9.201	
Epoch: 5	Cost: 0.041		Accuracy: 0.978		Time: 9.072	
Epoch: 6	Cost: 0.042		Accuracy: 0.981		Time: 8.980	
Epoch: 7	Cost: 0.039		Accuracy: 0.985		Time: 9.289	
Epoch: 8	Cost: 0.031		Accuracy: 0.987		Time: 9.250	
Epoch: 9	Cost: 0.023		Accuracy: 0.989		Time: 9.354	
Epoch: 10	Cost: 0.016		Accuracy: 0.990		Time: 9.342	
Epoch: 11	Cost: 0.012		Accuracy: 0.992		Time: 9.443	
Epoch: 12	Cost: 0.007		Accuracy: 0.993		Time: 9.747	
Epoch: 13	Cost: 0.006		Accuracy: 0.994		Time: 9.799	
Epoch: 14	Cost: 0.005		Accuracy: 0.996		Time: 10.270	
Epoch: 15	Cost: 0.003		Accuracy: 0.996		Time: 10.055	
Epoch: 16	Cost: 0.

In [4]:
    n_test = mnist.test.num_examples
    for x,y in zip(imgs_test, labels_test):
        x = np.array(x, ndmin=2).T
        y = np.array(y, ndmin=2).T
        z1 = np.dot(w1, x) + b1
        a1 = relu(z1)
        z2 = np.dot(w2, a1) + b2
        a2 = softmax(z2)
        error_t += int(np.argmax(a2) != np.argmax(y))
    print("Test Accuracy: {:.3f}\t".format(1 - error_t/n_test))

Test Accuracy: 0.972	


## Part e:  
The cost and training time of each epoch is list in previous cells along with the test accuracy. To save space, only the result from batch size 64 is printed out. I summarized my result of different batch size in the following table.

| batch size 	| cost of first epoch 	| average training time 	| test accuracy 	|
|:----------:	|:-------------------:	|:---------------------:	|:-------------:	|
|     16     	|        0.001        	|         10.022        	|     0.977     	|
|     64     	|        0.332        	|         9.482         	|     0.972     	|
|     256    	|        0.860        	|         9.064         	|     0.971     	|
|    1024    	|        0.943        	|         8.626         	|     0.951     	|

From the table you can see that, as the batch size grows, the cost of first epoch is getting bigger. Besides, I found out cost almost converge to 0 after 20 epochs except batch size of 1024. When the batch size is 1024, even after 20 epochs of training, the cost is still 0.187. However, when it comes to training time, larger batch size has the advantage. I think this is because for smaller batch size, it will need iterate more times in one epoch, thus slowing down the speed. Since the accuracy is closely related to cost, it's obvious smaller batch size is better in test accuracy. Therefore, we need to choose a good batch size in training process to balance training speed and model performance.

## Part f:  
Implement ADAM Optimizer and do the summary as part e

In [5]:
import numpy as np
import time
from tensorflow.examples.tutorials.mnist import input_data

mnist = input_data.read_data_sets("Mnist_data/", one_hot=True)
IMGS = mnist.train.images
LABS = mnist.train.labels
IMGS_Test = mnist.test.images
LABS_Test = mnist.test.labels

def softmax(z):
    return np.exp(z) / np.sum(np.exp(z))


def softmax_prime(z):
    return softmax(z) * (1 - softmax(z))

def relu(z):
    return np.maximum(z, 0)

# def relu_prime(z):
#     result = []
#     for t in z:
#         result.append(float(t > 0))
#     return np.array(result, ndmin=2).T

# This one is a lot faster
def relu_prime(z):
    x = np.array(z)
    return np.ceil((abs(x)+x)/2/100)

epochs = 20
lr = 0.2
batch_size = 1024
n = mnist.train.num_examples
total_batch = int(n/batch_size)
n_step = epochs * total_batch

beta_1 = 0.9
beta_2 = 0.999
epsilon = 0.00000001

n_input = 784
n_hidden = 100
n_output = 10

w1 = np.random.normal(0, 0.1 ,(n_hidden, n_input))
b1 = np.random.normal(0, 0.1 ,(n_hidden, 1))
w2 = np.random.normal(0, 0.1 ,(n_output, n_hidden))
b2 = np.random.normal(0, 0.1 ,(n_output, 1))

for j in range(epochs):
    time_start = time.time()
    error = 0
    error_t = 0
    cost = 0
    for k in range(0,n,batch_size):
        imgs = IMGS[k:k+batch_size]
        labels = LABS[k:k+batch_size]
        imgs_test = IMGS_Test
        labels_test = LABS_Test
        w1_gd = 0
        w2_gd = 0
        b1_gd = 0
        b2_gd = 0
        t = 0
        a2_m = 0
        a1_m = 0
        a2_v = 0
        a1_v = 0
        a2_m_b = 0
        a1_m_b = 0
        a2_v_b = 0
        a1_v_b = 0

        for x, y in zip(imgs, labels):
            x = np.array(x, ndmin=2).T
            y = np.array(y, ndmin=2).T
            z1 = np.dot(w1, x) + b1
            a1 = relu(z1)
            z2 = np.dot(w2, a1) + b2
            a2 = softmax(z2)

            d2 = a2 - y
            d1 = relu_prime(z1)*np.dot(w2.T, d2)

            w1_gd += np.dot(d1, x.T)
            w2_gd += np.dot(d2, a1.T)
            b1_gd += d1
            b2_gd += d2
            error += int(np.argmax(a2) != np.argmax(y))
            cost = np.sum(y*np.log(a2) + (1-y)*np.log(1-a2))
            # Adam
            t += 1
            a2_m = a2_m * beta_1 + (1-beta_1)*w2_gd
            a1_m = a1_m * beta_1 + (1-beta_1)*w1_gd
            a2_m_b = a2_m_b * beta_1 + (1-beta_1)*b2_gd
            a1_m_b = a1_m_b * beta_1 + (1-beta_1)*b1_gd
        
            a2_v = a2_v * beta_2 + (1-beta_2)* (w2_gd ** 2)
            a1_v = a1_v * beta_2 + (1-beta_2)* (w1_gd ** 2)
            a2_v_b = a2_v_b * beta_2 + (1-beta_2)* (b2_gd ** 2)
            a1_v_b = a1_v_b * beta_2 + (1-beta_2)* (b1_gd ** 2)
        
            a2_m_nor = a2_m / (1 - (beta_1 ** t))
            a1_m_nor = a1_m / (1 - (beta_1 ** t))
            a2_v_nor = a2_v / (1 - (beta_2 ** t))
            a1_v_nor = a1_v / (1 - (beta_2 ** t))
            a2_m_b_nor = a2_m_b / (1 - (beta_1 ** t))
            a1_m_b_nor = a1_m_b / (1 - (beta_1 ** t))
            a2_v_b_nor = a2_v_b / (1 - (beta_2 ** t))
            a1_v_b_nor = a1_v_b / (1 - (beta_2 ** t))        
        w2 -= (a2_m_nor / (np.sqrt(a2_v_nor) + epsilon)) / batch_size
        w1 -= (a1_m_nor / (np.sqrt(a1_v_nor) + epsilon)) / batch_size
        b2 -= (a2_m_b_nor / (np.sqrt(a2_v_b_nor) + epsilon)) / batch_size
        b1 -= (a1_m_b_nor / (np.sqrt(a1_v_b_nor) + epsilon)) / batch_size
        
#         # Gradient Descent
#         w1 -= lr * w1_gd / batch_size
#         w2 -= lr * w2_gd / batch_size
#         b1 -= lr * b1_gd / batch_size
#         b2 -= lr * b2_gd / batch_size
    time_end = time.time()
    print("Epoch: "+str(j+1)+"\tCost: {:.3f}\t".format(-cost)+"\tAccuracy: {:.3f}\t".format(1 - error/n)+
          "\tTime: {:.3f}\t".format(time_end-time_start))

Extracting Mnist_data/train-images-idx3-ubyte.gz
Extracting Mnist_data/train-labels-idx1-ubyte.gz
Extracting Mnist_data/t10k-images-idx3-ubyte.gz
Extracting Mnist_data/t10k-labels-idx1-ubyte.gz
Epoch: 1	Cost: 0.589		Accuracy: 0.772		Time: 33.853	
Epoch: 2	Cost: 0.384		Accuracy: 0.904		Time: 34.114	
Epoch: 3	Cost: 0.284		Accuracy: 0.921		Time: 35.328	
Epoch: 4	Cost: 0.240		Accuracy: 0.932		Time: 34.746	
Epoch: 5	Cost: 0.193		Accuracy: 0.939		Time: 35.626	
Epoch: 6	Cost: 0.159		Accuracy: 0.945		Time: 46.546	
Epoch: 7	Cost: 0.135		Accuracy: 0.950		Time: 38.990	
Epoch: 8	Cost: 0.120		Accuracy: 0.954		Time: 36.597	
Epoch: 9	Cost: 0.102		Accuracy: 0.957		Time: 36.108	
Epoch: 10	Cost: 0.088		Accuracy: 0.960		Time: 35.580	
Epoch: 11	Cost: 0.075		Accuracy: 0.962		Time: 35.199	
Epoch: 12	Cost: 0.061		Accuracy: 0.964		Time: 35.048	
Epoch: 13	Cost: 0.048		Accuracy: 0.966		Time: 35.089	
Epoch: 14	Cost: 0.038		Accuracy: 0.967		Time: 35.425	
Epoch: 15	Cost: 0.030		Accuracy: 0.969		Time: 38.514	
Epoch

In [6]:
    n_test = mnist.test.num_examples
    for x,y in zip(imgs_test, labels_test):
        x = np.array(x, ndmin=2).T
        y = np.array(y, ndmin=2).T
        z1 = np.dot(w1, x) + b1
        a1 = relu(z1)
        z2 = np.dot(w2, a1) + b2
        a2 = softmax(z2)
        error_t += int(np.argmax(a2) != np.argmax(y))
    print("Test Accuracy: {:.3f}\t".format(1 - error_t/n_test))

Test Accuracy: 0.966	


In order to show the advantage of ADAM Optimizer compare to Gradient Descent, this time I print out the result for batch size of 1024 because it performs the worst in last part. As we can see from the following table, in the case of 1024 batch size, the cost of first epoch drop to 0.589 compare to 0.943 in part e, and the test accuracy increase to 0.966. Also, we can notice that after 20 epoches of training, cost converge to 0.008 which is way better than 0.187 when we use gradient descent. Another huge change is the average training time for one epoch, it goes up to around 40 seconds in this case mainly because much more computations are brought by ADAM Optimizer. I'm still using CPU training in problem 1, so you can see although the pattern of training time is the same as last part, which goes down when increase batch size, the last one is actually bigger than size of 256. I think this might because after all these calculation, the CPU is too hot, and leads to CPU clock frequency reduction. In conclusion, ADAM improve the performance but is more complex in computation.

| batch size 	| cost of first epoch 	| average training time 	| test accuracy 	|
|:----------:	|:-------------------:	|:---------------------:	|:-------------:	|
|     16     	|        0.000        	|         42.389        	|     0.977     	|
|     64     	|        0.000        	|         38.437        	|     0.977     	|
|     256    	|        0.003        	|         35.212        	|     0.973     	|
|    1024    	|        0.589        	|         36.308        	|     0.966     	|

## Part g:  
TensorFlow implementation.

In [13]:
import time
import tensorflow as tf
old_v = tf.logging.get_verbosity()
tf.logging.set_verbosity(tf.logging.ERROR)

from tensorflow.examples.tutorials.mnist import input_data
#get mnist data, with one_hot encoding
mnist = input_data.read_data_sets("MNIST_data/",one_hot=True)
#suppress warnings
tf.logging.set_verbosity(old_v)

#learning rate
lr = 0.2
#number of traning epochs
epochs = 20
#number of batch_size
batch_size = 64
total_batch = int(mnist.train.num_examples/batch_size)
num_steps = epochs * total_batch

#network parameters
n_hidden_1 = 100
# n_hidden_2 = 200
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1],stddev=0.1),name='W1'),
    # 'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2],stddev=0.1),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_1, num_classes],stddev=0.1),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    # 'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

#define a neural net model
def neural_net(x):
    layer_1_out = tf.nn.relu(tf.add(tf.matmul(x,weights['W1']),biases['b1']))
    # layer_2_out = tf.nn.relu(tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2']))
    out = tf.add(tf.matmul(layer_1_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)
Y_hat = tf.nn.softmax(logits)

#define cost
cost = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='cost')
#define optimizer
optimizer = tf.train.GradientDescentOptimizer(learning_rate=lr)
train_op = optimizer.minimize(cost)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(Y_hat,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    time_start = time.time()
    for i in range(num_steps):
        # fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        # run optimization
        _, loss = sess.run([train_op, cost], feed_dict={X: batch_x, Y: batch_y})
        if i % total_batch == 0:
            time_end = time.time()
            acc = sess.run(accuracy, feed_dict={X: batch_x, Y: batch_y})
            print("Epoch " + str(i/total_batch+1) +"\tCost: {:.3f}\t".format(loss)+", Accuracy= {:.3f}".format(acc)+
                  "\tTime: {:.3f}\t".format(time_end-time_start))
            time_start = time.time()

    print("Training finished!")

    print("Testing ACcuracy:", sess.run(accuracy, feed_dict={X: mnist.test.images, Y: mnist.test.labels}))

Extracting MNIST_data/train-images-idx3-ubyte.gz
Extracting MNIST_data/train-labels-idx1-ubyte.gz
Extracting MNIST_data/t10k-images-idx3-ubyte.gz
Extracting MNIST_data/t10k-labels-idx1-ubyte.gz
Epoch 1.0	Cost: 2.491	, Accuracy= 0.297	Time: 0.187	
Epoch 2.0	Cost: 0.076	, Accuracy= 1.000	Time: 0.845	
Epoch 3.0	Cost: 0.128	, Accuracy= 1.000	Time: 0.824	
Epoch 4.0	Cost: 0.158	, Accuracy= 0.984	Time: 0.695	
Epoch 5.0	Cost: 0.119	, Accuracy= 0.984	Time: 0.839	
Epoch 6.0	Cost: 0.102	, Accuracy= 1.000	Time: 0.851	
Epoch 7.0	Cost: 0.075	, Accuracy= 1.000	Time: 0.855	
Epoch 8.0	Cost: 0.094	, Accuracy= 0.984	Time: 0.863	
Epoch 9.0	Cost: 0.109	, Accuracy= 1.000	Time: 0.874	
Epoch 10.0	Cost: 0.002	, Accuracy= 1.000	Time: 0.890	
Epoch 11.0	Cost: 0.010	, Accuracy= 1.000	Time: 0.870	
Epoch 12.0	Cost: 0.030	, Accuracy= 1.000	Time: 0.893	
Epoch 13.0	Cost: 0.060	, Accuracy= 1.000	Time: 0.883	
Epoch 14.0	Cost: 0.012	, Accuracy= 1.000	Time: 0.890	
Epoch 15.0	Cost: 0.025	, Accuracy= 1.000	Time: 0.924	
Epoch

As we can see from tensorflow implementation, the training time for 1 epoch is 10 times faster than numpy implementation. Besides, it also converges much faster in tensorflow. I think this might because a lot of optimizations have been performed in tensorflow, and thus make it fast to train and converge. The test accuracy is also 0.977, which is the same as my numpy implementation. This means at least my numpy implementation works and the upper limit accuracy of this structure is around 0.977. The following table is the table for different batch size.

| batch size 	| cost of first epoch 	| average training time 	| test accuracy 	|
|:----------:	|:-------------------:	|:---------------------:	|:-------------:	|
|     16     	|        0.030        	|         2.674         	|     0.977     	|
|     64     	|        0.057        	|         0.867         	|     0.977     	|
|     256    	|        0.326        	|         0.986         	|     0.973     	|
|    1024    	|        0.510        	|         0.943         	|     0.948     	|